pkg/partitionprune/filter.go

```diff
@@ -0,0 +1,794 @@
+			res[i] = int(pt.Position)
+		}
+		return res, nil
+	}
+	switch metadata.Method {
+	case partition.PartitionMethod_Range:
+		return rangeFilter(proc, filters, metadata)
+	case partition.PartitionMethod_Hash,
+		partition.PartitionMethod_Key:
+		return hashFilter(proc, filters, metadata)
+	case partition.PartitionMethod_List:
+		return listFilter(proc, filters, metadata)
+	}
+	return nil, nil
+}
+
+func hashFilter(
+	proc *process.Process,
+	filters []*plan.Expr,
+	metadata partition.PartitionMetadata,
+) ([]int, error) {
+	colPosition := mustGetColPosition(metadata.Partitions[0].Expr)
+	tm := map[int]struct{}{}
+	for _, expr := range filters {
+		targets, ok, err := hashFilterExpr(proc, colPosition, expr, metadata)
+		if err != nil {
+			return nil, err
+		}
+		if !ok {
+			res := make([]int, len(metadata.Partitions))
+			for i, pt := range metadata.Partitions {
+				res[i] = int(pt.Position)
+			}
+			return res, nil
+		}
+		for _, target := range targets {
+			tm[target] = struct{}{}
+		}
+	}
+	res := make([]int, 0, len(tm))
+	for target := range tm {
+		res = append(res, target)
+	}
+	return res, nil
+}
+
+func hashFilterExpr(
+	proc *process.Process,
+	colPosition int32,
+	expr *plan.Expr,
+	metadata partition.PartitionMetadata,
+) ([]int, bool, error) {
+	exprs := make([]*plan.Expr, len(metadata.Partitions))
+	for i, pt := range metadata.Partitions {
+		exprs[i] = p.DeepCopyExpr(pt.Expr)
+	}
+	switch exprImpl := expr.Expr.(type) {
+	case *plan.Expr_F:
+		switch exprImpl.F.Func.ObjName {
+		case "or":
+			left, can, err := hashFilterExpr(proc, colPosition, exprImpl.F.Args[0], metadata)
+			if err != nil {
+				return nil, false, err
+			}
+			if !can {
+				return nil, false, nil
+			}
+			right, can, err := hashFilterExpr(proc, colPosition, exprImpl.F.Args[1], metadata)
+			if err != nil {
+				return nil, false, err
+			}
+			if !can {
+				return nil, false, nil
+			}
+			sort.Ints(left)
+			sort.Ints(right)
+			return mergeSortedSlices(left, right), true, nil
+
+		case "and":
+			left, can, err := hashFilterExpr(proc, colPosition, exprImpl.F.Args[0], metadata)
+			if err != nil {
+				return nil, false, err
+			}
+			if !can {
+				return nil, false, nil
+			}
+			right, can, err := hashFilterExpr(proc, colPosition, exprImpl.F.Args[1], metadata)
+			if err != nil {
+				return nil, false, err
+			}
+			if !can {
+				return nil, false, nil
+			}
+			sort.Ints(left)
+			sort.Ints(right)
+			return intersectSortedSlices(left, right), true, nil
+		case "=":
+			left, ok := exprImpl.F.Args[0].Expr.(*plan.Expr_Col)
+			if !ok {
+				return nil, false, nil
+			}
+			if left.Col.ColPos != colPosition {
+				return nil, false, nil
+			}
+			for i := range exprs {
+				mustReplaceCol(exprs[i], exprImpl.F.Args[1])
+			}
+			targets, err := filterResult(proc, exprs, metadata)
+			if err != nil {
+				return nil, false, err
+			}
+			return targets, true, nil
+		}
+	}
+	return nil, false, nil
+}
+
+func rangeFilter(
```

Comment:
Same question of the loop.   Is this AND or OR?   Should it be AND instead of OR?